### PR TITLE
[i468] - sdapi_ingest_scripts

### PR DESCRIPTION
**Pull request by ShanaLMoore**
# Summary 

This PR has evolved from a temporary prod deploy branch to ingest sdapi data, to code we might need later. Instead of throwing this code away, we are storing it under a directory called sdapi_ingest_scripts for possible future use when we import ADL data. That way we can merge it into main but keep it isolated from the rest of the code base. 

ref: https://github.com/scientist-softserv/adventist-dl/issues/468

To run the jobs, from the cli...

```
require_relative 'path/to/my_custom_job' 
MyCustomJob.perform_later
```

## old comments: 

~~this branch includes work from multiple PRs in case for SDAPI related production deploys, without unapproved code from main. Most of this can be considered temporary - just meant to help get the SDAPI data in before the deadline vs actually merging this into main.~~

~~Regarding the rerun entry jobs... If they are considered useful, it should get contributed back to bulkax.~~ 

# 1. Upping memory limits (already merged into main)
https://github.com/scientist-softserv/adventist-dl/pull/447

# 2. Solr URI fix (needed to import Collection CSV w bulkrax) (already merged into main)

### STORY
Additional work may be needed to paginate the collection queries but this provides an initial quick solution.

Refs:

https://github.com/scientist-softserv/adventist-dl/issues/445

# 3. Rerun errored entries job (contribute back to bulkrax?)
https://github.com/scientist-softserv/adventist-dl/pull/440

```
RerunErroredEntriesForImporterJob.perform_later(importer_id: ##)
```

# 4. Collection/record relationships (too specific to Adventist but a similar concept could be appreciated in bulkrax)
ref: https://github.com/scientist-softserv/adventist-dl/pull/455

### Story

This creates a service to create relationships between works and collections based on a spreadsheet. A separate job is submitted for each relationship.

Call on rails console... switch to 'sdapi' tenant, then run `AddWorksToCollections.call`
Refs: 

- #444

# 5. Destroy Collections job (discard after sdapi?)
ref: https://github.com/scientist-softserv/adventist-dl/pull/452

```
BatchDestroyCollectionsJob.perform_later
```
# UpdateCollectionMembership Job
ref: #467 

### NOTES

Collection import csv: 
[SDAPI_periodical_collections_updated.csv](https://github.com/scientist-softserv/adventist-dl/files/11774694/SDAPI_periodical_collections_updated.csv)

### CODE REVIEW

File: app/jobs/rerun_entry_job.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  ##
+  # @param entry [Entry Object] the entry to run
+  def perform(bulkrax_entry: entry, importer_run: last_run)
+    logger = Rails.logger
+    logger.info("Submitting re-import for #{bulkrax_entry.class} ID=#{bulkrax_entry.id}")
+
+    bulkrax_entry.build
+    bulkrax_entry.save
+
+    if entry.status == "Complete"
```
Comment (ShanaLMoore) [inline app/jobs/rerun_entry_job.rb:20]:
```suggestion
    if bulkrax_entry.status == "Complete"
```

File: app/jobs/rerun_entry_job.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+# This class is responsible for running a single bulkrax entry.
+#
+# @note This class is starting it's life in the Adventist repository but is being designed so that
+#       we could copy it rather easily into Bulkrax.
+#
+# separating this into a job will allow us to run multiple jobs concurrently in the background
+class RerunEntryJob < ApplicationJob
+  ##
+  # @param entry [Entry Object] the entry to run
+  def perform(entry_class_name:, entry_id:, importer_run: last_run)
```
Comment (ShanaLMoore) [inline app/jobs/rerun_entry_job.rb:13]: try passing importer_run_id instead

File: app/jobs/rerun_entry_job.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  # @param entry [Entry Object] the entry to run
+  def perform(entry_class_name:, entry_id:)
+    bulkrax_entry = entry_class_name.constantize.find(entry_id)
+    Rails.logger.info("Submitting re-import for #{bulkrax_entry.class} ID=#{bulkrax_entry.id}")
+
+    begin
+      bulkrax_entry.build
+      bulkrax_entry.save
+    rescue StandardError => e
+      logger.error("Error re-submitting entry for #{bulkrax_entry.class} ID=#{bulkrax_entry.id}: #{e.message}")
```
Comment (kirkkwang) [inline app/jobs/rerun_entry_job.rb:23]: not sure if it needs it but maybe this?

```suggestion
      Rails.logger.error("Error re-submitting entry for #{bulkrax_entry.class} ID=#{bulkrax_entry.id}: #{e.message}")
```

**Comment by ShanaLMoore** [comment]
For the record, I am going to cherry-pick the changes here into the rerun branch to avoid the changes from main.

File: app/jobs/add_works_to_collections_job.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+    collection.try(:reindex_extent=, Hyrax::Adapters::NestingIndexAdapter::LIMITED_REINDEX)
+    work.member_of_collections << collection
+    work.save!
+    Rails.logger.info("🦄🪺👩‍👩‍👧‍👦 #{record_data[:identifier]} processed")
+  rescue StandardError => e
+    notice = "😈😈😈 ERROR: unable to create relationship for #{record_data[:identifier]}"
+    Rails.logger.error(notice)
+    Rails.logger.error(e.message)
+    File.open('lib/data/add_works_to_collection_errors.txt', 'a') do |f|
```
Comment (ShanaLMoore) [inline app/jobs/add_works_to_collections_job.rb:18]: This needs to be saved to a persisted volume director: 

image

File: app/jobs/add_works_to_collections_job.rb [hunk 5]
```diff
@@ -7,6 +7,9 @@ def perform(record_data:)
     work = record_data[:model].constantize.where(identifier: record_data[:identifier]).first
     collection = Collection.where(identifier: record_data[:parents]).first
 
+    return if !work.present? && !collection.present?
+    return if work.member_of_collection_ids.include?(collection.id)
+
```
Comment (ShanaLMoore) [inline app/jobs/add_works_to_collections_job.rb:12]: confirmed that this is working. When the id isn't included it created the relationship: 

image


image

File: sdapi_ingest_scripts/configs/raven.rb [hunk 6]
```diff
@@ -0,0 +1,6 @@
+# frozen_string_literal: true
+
+Raven.configure do |config|
+  config.dsn = ENV['SENTRY_DSN']
+  config.current_environment = Rails.env
+end
```
Comment (laritakr) [inline sdapi_ingest_scripts/configs/raven.rb:6]: Wondering if this might be useful to keep? Were these changes needed for the jobs to pick up errors in addition to the change in good_job.rb? 

Also, everything reports under production in Sentry as it currently is set up, so this should give us the environment breakdown. Or maybe check with April to see if there are other changes that could be made so it works right without it?

File: sdapi_ingest_scripts/rerun_entry_job.rb [hunk 7]
```diff
@@ -0,0 +1,28 @@
+# frozen_string_literal: true
```
Comment (laritakr) [inline sdapi_ingest_scripts/rerun_entry_job.rb:1]: Should this not be inside the one_off subdirectory?

File: sdapi_ingest_scripts/rerun_errored_entries_for_importer_job.rb [hunk 8]
```diff
@@ -0,0 +1,89 @@
+# frozen_string_literal: true
```
Comment (laritakr) [inline sdapi_ingest_scripts/rerun_errored_entries_for_importer_job.rb:1]: This one as well?

File: sdapi_ingest_scripts/configs/good_job.rb [hunk 9]
```diff
@@ -1,9 +1,11 @@
 
+# frozen_string_literal: true
+
 Rails.application.configure do
   # Configure options individually...
   config.good_job.preserve_job_records = true
   config.good_job.retry_on_unhandled_error = false
-  config.good_job.on_thread_error = -> (exception) { Sentry.capture_exception(exception) }
+  config.good_job.on_thread_error = ->(exception) { Raven.capture_exception(exception) }
```
Comment (laritakr) [inline sdapi_ingest_scripts/configs/good_job.rb:8]: I think we want to keep this change. It's needed for jobs to pick up errors.

**Review by laritakr — CHANGES_REQUESTED** [review]

File: sdapi_ingest_scripts/configs/raven.rb (hunk 6 above)
Comment (ShanaLMoore) [reply sdapi_ingest_scripts/configs/raven.rb:6]: Ok, I'll take some time to find out how sentry works. Rob said we shouldn't need this file at all so I wasn't sure about adding it directly to main.

File: config/initializers/raven.rb [hunk 10]
```diff
@@ -0,0 +1,6 @@
+# frozen_string_literal: true
```
Comment (ShanaLMoore) [inline config/initializers/raven.rb:1]: TODO: set [SENTRY_ENVIRONMENT](https://github.com/scientist-softserv/atla-hyku/blob/2e938ea267a1719fa6f01b7277a65f7ae3bc0904/ops/production-deploy.tmpl.yaml#L179C11-L179C29) variable instead of adding an initializer file.

File: sdapi_ingest_scripts/one_off/jobs/add_works_to_collections_job.rb [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+
+require 'csv'
+
+class AddWorksToCollectionsJob < ApplicationJob
+  queue_as :relationships
+
+  retry_on StandardError, attempts: 0
+
+  def perform
+    csv_path = "lib/data/articles_parent_child_relationships_final.csv"
```
Comment (laritakr) [inline sdapi_ingest_scripts/one_off/jobs/add_works_to_collections_job.rb:11]: Why not put the csv data into `sdapi_ingest_scripts/data` and adjust the path?

Comment (ShanaLMoore) [reply sdapi_ingest_scripts/one_off/jobs/add_works_to_collections_job.rb:11]: sure, can do
